pkg/spdk/engine.go

```diff
@@ -1229,6 +1333,399 @@ func (e *Engine) validateAndUpdateReplicaNvme(replicaName string, bdev *spdktype
 	return e.ReplicaStatusMap[replicaName].Mode, nil
 }
 
+// This method performs an online volume expansion for the Longhorn Engine using SPDK. It:
+// Expands underlying replica logical volumes (lvol)
+// Recreates the SPDK RAID bdev
+// Suspends and resumes frontend I/O as needed
+// Ensures cleanup and status update on failure
+func (e *Engine) Expand(spdkClient *spdkclient.Client, size uint64) (retErr error) {
+	e.Lock()
+	defer e.Unlock()
+
+	e.log.Info("Expanding engine")
+
+	if err := e.startExpansion(size); err != nil {
+		return errors.Wrap(err, "startExpansion failed")
+	}
+
+	var (
+		expanded  = false
+		engineErr error
+	)
+
+	// engineErr will be set when the engine failed to do any non-recoverable operations.
+	defer func() {
+		if retErr != nil {
+			e.log.WithError(engineErr).Errorf("Engine %s failed to expand", e.Name)
+			e.lastExpansionError = errors.Wrap(retErr, "engine failed to expand expansion").Error()
+			e.lastExpansionFailedAt = time.Now().UTC().Format(time.RFC3339Nano)
+		}
+
+		if engineErr != nil {
+			if e.State != types.InstanceStateError {
+				e.State = types.InstanceStateError
+			}
+			e.ErrorMsg = engineErr.Error()
+
+			if errUpdateLogger := e.log.UpdateLogger(logrus.Fields{
+				"replicaStatusMap": e.ReplicaStatusMap,
+			}); errUpdateLogger != nil {
+				e.log.WithError(errUpdateLogger).Warn("Failed to update logger with replica status map during engine creation")
+			}
+
+			e.lastExpansionError = errors.Wrap(retErr, "engine under non-recoverable operation").Error()
+			e.lastExpansionFailedAt = time.Now().UTC().Format(time.RFC3339Nano)
+		}
+
+		e.finishExpansion(expanded, size)
+	}()
+
+	// fetch all replica clients
+	replicaClients, err := e.getReplicaClients()
+	if err != nil {
+		retErr = err
+		return retErr
+	}
+	defer e.closeRplicaClients(replicaClients)
+	if e.isAnyReplicaUnderRebuilding(replicaClients) {
+		retErr = errors.New("expansion failed: at least one replica is under rebuilding")
+		return retErr
+	}
```

Comment:
rebuilding replica can be detected by `replicaStatus.Mode != types.ModeRW` above